barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_trace.cpp

```diff
@@ -2295,6 +2296,130 @@ uint32_t AvmTraceBuilder::read_slice_to_memory(uint8_t space_id,
+    std::vector<uint32_t> second_row_values = {};
+    std::vector<uint32_t> second_row_load = {
+        uint32_t(read_ind_ret_offset.val), function_selector_offset, ret_size, success_offset
+    };
+    for (uint32_t j = 0; j < second_row_load.size(); j++) {
+        // We just read and load to set up the constraints, we dont actually use these values for now.
+        auto mem_read = mem_trace_builder.read_and_load_from_memory(call_ptr,
+                                                                    clk,
+                                                                    register_order[j % register_order.size()],
+                                                                    second_row_load[j],
+                                                                    AvmMemoryTag::U32,
+                                                                    AvmMemoryTag::U0);
+        second_row_values.emplace_back(mem_read.val);
+    }
+    main_trace.push_back(Row{
+        .avm_main_clk = clk,
+        .avm_main_ia = second_row_values[0], /* ret_offset */
+        .avm_main_ib = second_row_values[1], /* function_selector_offset */
+        .avm_main_ic = second_row_values[2], /* ret_size */
+        .avm_main_id = second_row_values[3], /* success_offset */
+        .avm_main_ind_a = ret_offset,
+        .avm_main_ind_op_a = FF(1),
+        .avm_main_internal_return_ptr = FF(internal_return_ptr),
+        .avm_main_mem_idx_a = read_ind_ret_offset.val,
+        .avm_main_mem_idx_b = function_selector_offset,
+        .avm_main_mem_idx_c = ret_size,
+        .avm_main_mem_idx_d = success_offset,
+        .avm_main_mem_op_a = FF(1),
+        .avm_main_mem_op_b = FF(1),
+        .avm_main_mem_op_c = FF(1),
+        .avm_main_mem_op_d = FF(1),
+        .avm_main_pc = FF(pc),
+        .avm_main_r_in_tag = FF(static_cast<uint32_t>(AvmMemoryTag::U32)),
+    });
+    // NOTE: we do not constrain this here as it's behaviour will change fully once we have a full enqueued function
+    // call in one vm circuit
+    clk++;
+    // PLAN
+    // 1. We load and constrain all of the operands ( of which there are loads, (either 4 more registers or we send
+    // it across two rows (we might be able to get away with this - and break a shit load of other stuff)))
+    std::vector<uint32_t> write_operands = { ret_offset, success_offset };
+    // 2. Use the write slice to memory once we get the entire return data size from the hint
+    write_slice_to_memory(call_ptr,
+                          clk,
+                          uint32_t(read_ind_ret_offset.val),
+                          AvmMemoryTag::U0,
+                          AvmMemoryTag::FF,
+                          internal_return_ptr,
+                          execution_hints.returndata_hints.at(return_data_counter));
+    return_data_counter++;
+    clk++;
+    write_slice_to_memory(
```

Comment:
does this have to be a slice if it's just one value?

Reply:
Not really, but the helper function also sets up the main trace and memory trace rows under the hood, so it ends up being a bit cleanr than expanding and manually constructing the row entries.